modules/images/webp-uploads/load.php

```diff
@@ -554,11 +554,31 @@ function webp_uploads_img_tag_update_mime_type( $image, $context, $attachment_id
 		return $image;
 	}
 
-	$urls = $matches[0];
-	// TODO: Add a filterable option to change the selected mime type. See https://github.com/WordPress/performance/issues/187.
-	$target_mime = 'image/webp';
+	/**
+	 * Filters mime types that should be used to update all images in the content. The order of
+	 * mime types matters. The last mime type in the list will be used if it is supported by an image.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param array The list of mime types that can be used to update images in the content.
+	 */
+	$target_mimes = apply_filters( 'webp_uploads_content_image_mimes', array( 'image/jpeg', 'image/webp' ) );
+
+	$target_mime  = null;
+	$target_mimes = array_reverse( $target_mimes );
+	foreach ( $target_mimes as $mime ) {
```

Comment:
Introducnig a `for` loop instead a `foreach`  helps to avoid the need for a reverse operation, and provides a more natural flow.

```suggestion
    for ( $i = count( $target_mimes ) - 1; $i >= 0; $i-- ) {
      $mime = $target_mimes[ $i ];
```

Reply:
I would still prefer to see this work in the reverse order: the first available mime type image in the list would be used instead of the last available type.

Reply:
> I would still prefer to see this work in the reverse order: the first available mime type image in the list would be used instead of the last available type.

I see the benefit of that approach as well as is clear and more obvious by just looking at the order of the provided array.

Reply:
@mitogh, unfortunately, your `for` loop will stop working as soon as someone unsets a mime type in a filter hook:

```php
add_filter( 'webp_uploads_content_image_mimes', function ( $types ) {
    $types[] = 'image/avif';
    return $types;
}, 10 );

add_filter( 'webp_uploads_content_image_mimes', function ( $types ) {
    unset( $types[ array_search( 'image/webp', $types, true ) ] );
    return $types;
}, 15 );

...

for ( $i = count( $target_mimes ) - 1; $i >= 0; $i-- ) {
    $mime = $target_mimes[ $i ]; // --> "PHP Warning:  Undefined array key 1 ..." - because the $target_mimes is "array( 0 => 'image/jpeg', 2 => 'image/avif' )
}
```

Reply:
Yeah, we would need to call `array_values` to make that reliable, so then it's also an extra array function call, so probably not better than `array_reverse`.

Reply:
> unfortunately, your for loop will stop working as soon as someone unsets a mime type in a filter hook:

If an element is unset from an an array the `count` value would still return the number:

```php
// Count => 3
$types = array( 'image/jpeg', 'image/webp', 'image/avif');

unset( $types[ array_search( 'image/webp', $types, true ) ] );
// After unset
count( $types ) // Returns 2
// So the for loop will start from N - 1 in this case 1 and 0 indexes 
// so even when using the unset the for loop will still work as expected.
```

Not sure if I am missing something, thanks.

Reply:
@mitogh The count is correct still, but the indexes aren't. The `for` loop would for your example go from index `1` to `0`. However, because index `1` (`image/webp`) was removed from the array, `$target_mimes[1]` is undefined, causing a problem here. And it would skip `$target_mimes[2]` which is still there as `image/avif`.

Reply:
Woops, my bad!

In that case using iterators is another option.

```php
end( $target_mimes );
for ( $i = count( $target_mimes ) - 1; $i >= 0; $i-- ) {
    $mime = current( $target_mimes );
    prev($target_mimes)
}
```

Reply:
@mitogh why do you think that using the `for` loop and a bunch of iteration function calls are better than just the `foreach` loop and a single `array_reverse` call?

Reply:
> why do you think that using the for loop and a bunch of iteration function calls are better than just the foreach loop and a single array_reverse call?

Not necessarily better but most performant due the avoiding the need to call `array_reverse` which is:

- N more in time to reverse
- N more in space to store the reverse of operation.

When the `for` is always `N` due no additional space or operation is required, I know this might be just 3 at the most elements but any cumulative gain would help in the long run.

@eugene-manuilov

Reply:
@mitogh, this performance optimisation doesn't make any sense for an array with a few items because even if there a performance gain (which is arguable) it will be scarce.

If you want to improve performance, I would encourage you to offer a replacement for all those preg_match calls to find images in the content which has much more significant impact on performance, especially for big posts with much content.

As to complexity of the algorithm, you need to take into account that function calls are expensive in PHP (you can read about it [here](https://stackoverflow.com/questions/3691625/why-are-php-function-calls-so-expensive)). So, sometimes it is easier and better to call one function with N complexity to avoid running an algorithm that has 2N+2 function calls.

Reply:
> this performance optimisation doesn't make any sense for an array with a few items because even if there a performance gain (which is arguable) it will be scarce.

Still hard to see why does not make sense having a performance gain.

> If you want to improve performance, I would encourage you to offer a replacement for all those preg_match calls to find images in the content which has much more significant impact on performance, especially for big posts with much content.

Under that premise we might need to go back into the WP Core and update many of those used to insert the `srcset` and other attributes, I think we need to take a look into a more iterative approach rather than a more drastic one.

Overall my main point is if we can optimize by just replacing `foreach` and an `array_reverse` would go for it but happy to move the code as it is.

Reply:
> Overall my main point is if we can optimize by just replacing foreach and an array_reverse would go for it but happy to move the code as it is.

IMO readability here worth more than a performance optimisation that makes the script run faster by ~0.0001ms. I don't mind changing it if we have a consensus here. @felixarntz @adamsilverstein what do you think?

Reply:
I'd be in favor for keeping this as is - we're speaking about a micro optimization here, and it's unclear whether the iterator approach is actually faster than `array_reverse`. The code is more familiar to read the way it is now with the `foreach` loop on the reversed array.

We shouldn't block this issue because of this minor point, so I'd say let's go with what we have now.